### ISSUE TITLE
Parser segfaults on invalid expressions

### ISSUE THREAD

**Issue by philip-blakely** [AMReX-Codes/amrex#2231]
Our group is finding the amrex::Parser very useful for allowing users to put equations into settings files. However, to help users find errors, we would like to be able to catch any invalid expressions so we can point them to the line in the settings file and expression that failed to work.

Unfortunately there are some invalid expressions that crash the parser in a confusing way:
```
amrex::Parser prs;
prs.define("x.x");
amrex::ParserExecutor<0> parser = prs.compile<0>();
amrex::Real value = parser();
```
gives the error:
```
Unknown character .
syntax error
Segfault
```
which is not very helpful to the end-user.

Changing "x.x" to "xx" gives:
```
terminate called after throwing an instance of 'std::runtime_error'
what():  Assertion `t->i >= 0' failed, file "amrex/Src/Base/Parser/AMReX_Parser_Exe.cpp", line 47, Msg: "parser_compile: unknown variable"
```
which would be OK as an error because we can catch and handle the exception.

We have also identified a few other expressions that seg-fault, such as:
- ``1e.``
- ``cod(x)``
- ``cod(1)``
that might be more plausible typos in a settings file.

In summary, it would be helpful if all syntax errors in amrex::Parser led to an exception.

Let me know if you would like any more context or information for this issue.

**Comment by WeiqunZhang** [comment]
If we make the parser throw an exception and you can catch it like this, would it be helpful to you?
```
        try {
            prs.define("x.x");
        } catch (const std::runtime_error& e) {   
            amrex::AllPrint() << e.what() << "\n";
        }
```

```
Unknown character .  in Parser::define("x.x")
```

**Comment by WeiqunZhang** [comment]
Looks like there are three types of errors.

1. Unknown character.  Example: `x.x`.
2. Syntax error.  Example: `cod(x)`.
3. Unknown variable. Example: `x+3`, but x is not registered as a variable.

**Comment by WeiqunZhang** [comment]
Using #2233, `xx` without `xx` being registered as a variable will result in
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Unknown variable xx in Parser expression "xx"
```

`x.x` will result in
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Unknown character . in Parser expression "x.x"
```

`1e.` will result in
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  syntax error in Parser expression "1e."
```

Does the PR address your issue?

**Comment by philip-blakely** [comment]
Yes, I've just applied the PR to my local amrex checkout and our unit-tests now pass again instead of seg-faulting.

Thank you very much!
